### PR TITLE
Fix initial sync race

### PR DESCRIPTION
**Pull request by lkysow**
A race condition was occurring where we would delete services via the
`watchReapableServices` function if `Sync` wasn't called fast enough by
`resource.go`. This occurred because `watchReapableServices` makes a
call to the catalog and then deletes everything with a k8s tag that
isn't in its `s.services` map. If `Sync` isn't called, then this map
is empty and so it tries to delete everything.
I've solved this by blocking until `Sync` is first called.

Closes https://github.com/hashicorp/consul-k8s/pull/191

### CODE REVIEW

**Comment by lkysow** [comment]
@ashangit sorry I didn't see your PR https://github.com/hashicorp/consul-k8s/pull/191 before putting this one up. I think this PR accomplishes the same fix but instead of adding a new `GetAllServices` method it waits for `Sync` to be called. Does this look like it should work for you?

**Comment by ashangit** [comment]
thks @lkysow looks good to me

File: catalog/to-consul/syncer_test.go [hunk 1]
```diff
@@ -155,6 +155,45 @@ func TestConsulSyncer_reapService(t *testing.T) {
...
+	defer a.Stop()
+	client, err := api.NewClient(&api.Config{
+		Address: a.HTTPAddr,
+	})
+	require.NoError(t, err)
+	s, closer := testConsulSyncer(client)
+	defer closer()
+
+	// Create a service directly in Consul. Since it was created directly we
+	// expect it to be deleted but not until the initial sync is performed.
```
Comment (adilyse) [inline catalog/to-consul/syncer_test.go]:
```suggestion
	// Create a service directly in Consul. Since it was created on the 
	// synthetic sync node and has the sync-associated tag, we expect
	// it to be deleted but not until the initial sync is performed.
```

A service needs the sync tag and to be registered on the sync node or the sync process would ignore it.

File: catalog/to-consul/syncer_test.go [hunk 2]
```diff
@@ -155,6 +155,45 @@ func TestConsulSyncer_reapService(t *testing.T) {
...
+	svc := testRegistration(ConsulSyncNodeName, "baz", "default")
+	_, err = client.Catalog().Register(svc, nil)
+	require.NoError(t, err)
+
+	// We wait until the syncer has had the time to delete the service.
+	time.Sleep(200 * time.Millisecond)
```
Comment (adilyse) [inline catalog/to-consul/syncer_test.go]: It seems like we might run into edge cases having this value exactly equal the `SyncPeriod` in the test ConsulSyncer. What about setting up the ConsulSyncer with an even smaller interval?

Comment (lkysow) [reply catalog/to-consul/syncer_test.go]: Yes! Great idea. I didn't like this sleep but a retry loop doesn't work because it might run through before syncFull has run :D

File: catalog/to-consul/syncer_test.go [hunk 3]
```diff
@@ -155,6 +155,51 @@ func TestConsulSyncer_reapService(t *testing.T) {
...
+	t.Parallel()
+
+	a, err := testutil.NewTestServerT(t)
+	require.NoError(t, err)
+	defer a.Stop()
+	client, err := api.NewClient(&api.Config{
+		Address: a.HTTPAddr,
+	})
+	require.NoError(t, err)
+	s, closer := testConsulSyncer(client)
```
Comment (adilyse) [inline catalog/to-consul/syncer_test.go]: Suggestion:
```
s, closer := testConsulSyncerWithConfig(client, func(s *ConsulSyncer) {
		s.SyncPeriod = 5 * time.Millisecond
	})
```

Since the `testConsulSyncer` also `Runs` the syncer, the sync period change may not get picked up if it's made afterwards.

Comment (lkysow) [reply catalog/to-consul/syncer_test.go]: Good catch. PR coming.

Comment (lkysow) [reply catalog/to-consul/syncer_test.go]: https://github.com/hashicorp/consul-k8s/pull/227
